[Min Hong] iP

auhc99 opened the pull request:
# DukeMH

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukeMH frees your mind of having to remember things you need to do. It's,

* text-based
*  easy to learn
* ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://github.com/auhc99/ip).
2. run it
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---
If you Java programmer, you can use it to practice Java too. Here's the ```main``` method:

```java
public class Duke {
     public static void main(String[] args) {
        runDuke();
    }
}
```

Commit:
Add Gradle support

auhc99 committed:
Level 1. Greet, Echo, Exit

Implement a skeletal version of Duke that starts by greeting the user, simply echos commands entered by the user, and exits when the user types bye.

auhc99 committed:
Level 2. Add, List

Add the ability to store whatever text entered by the user and display them back to the user when requested.

auhc99 committed:
Level 3. Mark as Done

Add the ability to mark tasks as done.

auhc99 committed:
Level 4. ToDos, Events, Deadlines

Add support for tracking three types of tasks:
1. ToDos: tasks without any date/time attached to it e.g., visit new theme park
2. Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
3. Events: tasks that start at a specific time and ends at a specific time e.g., team project meeting on 2/10/2019 2-4pm

auhc99 committed:
Automated Text UI Testing

Use the input/output redirection technique to semi-automate the testing of Duke.

auhc99 committed:
Level 5. Handle Errors

Teach Duke to deal with errors such as incorrect inputs entered by the user.

auhc99 committed:
Level 6. Delete

Add support for deleting tasks from the list.

auhc99 committed:
Fixed Automated Testing

Fixed EXPECTED.TXT and minor whitespace changes in Duke.java

auhc99 committed:
Add ability to save list

Level 7. Save
Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when Duke starts up. You may hard-code the file name and location e.g., [project_root]/data/duke.txt

auhc99 committed:
Add LocalDateTime

Level 8. Dates and Times
Teach Duke how to understand dates and times. For example, if the command is deadline return book /by 2/12/2019 1800, Duke should understand 2/12/2019 1800 as 2nd of December 2019, 6pm, instead of treating it as just a String.

auhc99 committed:
Merge branch 'branch-Level-8'

auhc99 committed:
Fix Saving Bug

auhc99 committed:
Refactor Code

Refactor the code to extract out closely related code as classes.

auhc99 committed:
Package code

Organize the classes into suitable java packages.

auhc99 committed:
Add JUnit Tests

Add JUnit tests to test the behavior of the code

auhc99 committed:
Add Manifest

auhc99 committed:
Add JavaDoc comments

auhc99 committed:
Tweak code to comply with Java coding standards

auhc99 committed:
Add find feature

auhc99 committed:
Merge branch 'branch-A-CodingStandard'

auhc99 committed:
Merge branch 'branch-Level-9'

auhc99 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

auhc99 committed:
Add Gradle support

yyhangz commented on src/main/java/duke/DukeList.java:
```diff
@@ -0,0 +1,110 @@
+     * Add a new task.
+     *
+     * @param task Task to be added.
+     */
```
The JavaDoc opening sentence should start with ```Adds... ``` instead of ```Add...```.

yyhangz commented on src/main/java/duke/DukeList.java:
```diff
@@ -0,0 +1,110 @@
+     * Delete a task.
+     *
+     * @param i Index of the task to be deleted.
+     */
```
The Javadoc opening sentence should start with ```Deletes... ``` instead of ```Delete...```.

yyhangz reviewed (commented):
Apart from the 2 sets of JavaDoc comments that I have commented on, there is really not much else to fault. I think you did a really good job so far, not just in terms of complying with the coding standard guidelines, but also the quality of the code itself (I really like the way you made use of abstract classes and inheritance in general; it makes it very easy to understand your code)

auhc99 committed:
Add GUI

tsm1820 commented on src/main/java/duke/DukeList.java:
```diff
@@ -0,0 +1,110 @@
+     * @param i Index of the task.
+     * @return Task.
+     */
+    public Task get(int i) {
```
Hi, overall you seemed to have followed the coding standards pretty well, so good job! Perhaps you could rename this method to getTask() to make it clearer what exactly is this method returning?

tsm1820 commented on src/main/java/duke/DukeList.java:
```diff
@@ -0,0 +1,110 @@
+     *
+     * @param task Task to be added.
+     */
+    public void add(Task task) {
```
Again same thing haha, rename it to addTask()?

tsm1820 commented on src/main/java/duke/DukeList.java:
```diff
@@ -0,0 +1,110 @@
+     *
+     * @param i Index of the task to be deleted.
+     */
+    public void delete(int i) {
```
Same, consider renaming it to deleteTask()?

tsm1820 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,235 @@
+     * @throws CorruptedFileException If file line is invalid.
+     */
+    public static Task parseFileLine(String line) throws CorruptedFileException {
+        String[] splitted = line.split("%,", 4);
```
Could perhaps rename this variable to specify what exactly are you splitting?

tsm1820 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,235 @@
+        String dateTime;
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/M/yyyy HHmm");
+        Task task;
+        switch (type) {
```
Nice use of the switch statement here

tsm1820 reviewed (commented):
LGTM. Your coding standard and code quality both seem up to the mark, so good job! Just some nits to fix.

auhc99 committed:
Add assertions

Add assertions to various points in the code

There are several assumptions that should hold at various points in the
code. These assumptions are critical for the program to run smoothly.

Adding assertions can detect possible bugs in the code because when an
assertion error is thrown, it indicates that the code has violated an
assumption about how the code should behave.

This commit also contains a bug fix for when the program loads a task
that was marked done from the save file, the task is added into the
list without being marked done.

auhc99 committed:
Update DukeList

Update DukeList

Updated DukeList to comply with checkstyle standards.

auhc99 committed:
Refactor Code

Refactor Code

Command classes (e.g DeadlineCommand, DeleteCommand, DoneCommand)
and the DialogBox class have some code fragment that can be grouped
together.

Those fragments can be extracted and turned into a method whose name
explains the purpose of the method.

Extracting the code fragments can improve the program's internal
structure, improving the code quality and making it easier for future
developers to read, understand and modify the code.

Let's extract the code fragments that can be grouped together into
another method, stringifyMessage for Command classes, initialiseLoader
and setDisplay for DialogBox.

auhc99 committed:
Merge pull request #1 from auhc99/branch-A-Assertions

Branch a assertions

auhc99 committed:
Merge branch 'master' into branch-A-CodeQuality

auhc99 committed:
Merge pull request #2 from auhc99/branch-A-CodeQuality

Branch a code quality

auhc99 committed:
Set up CI

auhc99 committed:
Add help command

Add help command

There are currently no commands to help users with the program.

Adding a help command can help new users learn about the many commands
available and their usage.

Let's add a help command that when triggered, display a list of
available commands and their usage.

auhc99 committed:
Rename methods

auhc99 committed:
Update .gitignore

auhc99 committed:
Improve GUI

auhc99 committed:
Add User Guide

CheahYan commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,30 @@
+
+import duke.DukeList;
+import duke.task.Deadline;
+
+public class DeadlineCommand extends DukeCommand {
+    private final Deadline task;
+
+    /**
+     * Public constructor to create a DoneCommand.
+     *
+     * @param task Deadline to be added.
+     */
+    public DeadlineCommand(Deadline task) {
+        super();
+        this.task = task;
+    }
+
+    @Override
+    public String run(DukeList list) {
+        list.addTask(task);
+        return stringifyMessage(list.size());
+    }
+
+    private String stringifyMessage(int listSize) {
+        return "Got it. I've added this task:\n"
+                + task.toString() + "\n"
+                + "Now you have " + listSize + " tasks in the list";
```
Perhaps you could add header comments for your methods and class like what you did for DeadlineCommand constructor. Otherwise, the code is easy to read and understand!

CheahYan commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,36 @@
+import duke.DukeList;
+import duke.exception.InvalidArgumentsException;
+import duke.task.Task;
+
+public class DeleteCommand extends DukeCommand {
+    private final int id;
+
+    /**
+     * Public constructor to create a DeleteCommand.
+     *
+     * @param id Id of task.
+     */
+    public DeleteCommand(int id) {
+        super();
+        this.id = id;
+    }
+
+    @Override
+    public String run(DukeList list) throws InvalidArgumentsException {
+        try {
+            Task task = list.getTask(id);
+            list.deleteTask(id);
+            return stringifyMessage(task.toString(), list.size());
+        } catch (IndexOutOfBoundsException e) {
+            throw new InvalidArgumentsException();
+        }
+    }
+
+    private String stringifyMessage(String taskString, int listSize) {
+        return "Noted. I've removed this task:\n"
+                + taskString + "\n"
+                + "Now you have " + listSize + " tasks in the list";
```
Same for here, you could add some header comments for your class and methods to improve readability.

CheahYan commented on src/main/java/duke/command/HelpCommand.java:
```diff
@@ -0,0 +1,24 @@
+
+import duke.DukeList;
+
+public class HelpCommand extends DukeCommand {
+    private static final String HELP_MESSAGE = "Here are the available commands:\n"
+            + "- Use \"help\" to display the list of commands.\n"
+            + "- Use \"list\" to display the list of tasks.\n"
+            + "- Use \"todo {description}\" to add a todo with a description.\n"
+            + "- Use \"event {description} /at {d/M/yyyy HHmm}\" to add an event with description and dateTime.\n"
+            + "- Use \"dateline {description} /by {d/M/yyyy HHmm}\" to add a deadline with description and dateTime.\n"
+            + "- Use \"done {n}\" to mark task number n done.\n"
+            + "- Use \"delete {n}\" to delete task number n.\n"
+            + "- Use \"find {keyWord}\" to find tasks with matching keywords.\n"
+            + "- Use \"bye\" to exit the program.\n";
+
+    public HelpCommand() {
+    }
+
+    @Override
+    public String run(DukeList list) {
+        return HELP_MESSAGE;
+    }
+}
```
Goob Job! It is very easy to read and understand your code. It looks very clean.

CheahYan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,243 @@
+            break;
+        case "D":
+            input = "deadline " + splitLine[2] + " /by " + splitLine[3];
+            type = TaskType.DEADLINE;
+            break;
+        case "E":
+            input = "event " + splitLine[2] + " /at " + splitLine[3];
+            type = TaskType.EVENT;
+            break;
+        default:
+            assert false : "Invalid task type";
+            throw new CorruptedFileException();
+        }
+        try {
+            Task task = parseTask(input, type);
+            if (isDone) {
+                task.markDone();
+            }
+            return task;
+        } catch (InvalidArgumentsException | InvalidTaskException e) {
+            throw new CorruptedFileException();
+        }
+    }
+
+    /**
+     * Returns the Task parsed from input.
+     *
+     * @param input Input to be parsed.
+     * @param type Type of Task.
+     * @return Task parsed.
+     * @throws InvalidArgumentsException If arguments are missing or invalid.
+     * @throws InvalidTaskException If input contains unknown commands.
+     */
+    private static Task parseTask(String input, TaskType type) throws InvalidArgumentsException, InvalidTaskException {
+        int descriptionEnd;
+        String description;
+        String dateTime;
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/M/yyyy HHmm");
+        Task task;
+        switch (type) {
```
Good Job! It is very easy to read your code as the flow of the code is separated very logically and code is abstracted nicely into methods.

CheahYan reviewed (commented):
Overall, code is very easy to read and logic is easy to follow. It will be even better if there are header comments for some of the classes and methods as it would greatly improve readability. LGTM!!

auhc99 committed:
Set theme jekyll-theme-cayman

auhc99 committed:
Update build.gradle

auhc99 committed:
Refactor Code

Refactor code to remove long methods
